### PR TITLE
Add supporting files for JSInterop packaging

### PR DESCRIPTION
**Pull request by vishnureddy17**
This PR implements the changes discussed in #18

### CODE REVIEW

**Comment by vishnureddy17** [comment]
Do we need a copyright/license header in the JS and TS files?

**Review by vishnureddy17 — COMMENTED** [review]
Do we need a copyright/license header in the JS and TS files?

**Review by rido-min — CHANGES_REQUESTED** [review]
I'd move the JSInterop project to the javascript folder
Not sure if we want to produce a former npm package

Why we need to update the scripts? build can be done with standard dotnet cli, and we are not planning to pack this project

**Comment by jrdouceur** [comment]
> I'd move the JSInterop project to the javascript folder

It is dotnet code; shouldn't it go in the dotnet folder?

> Why we need to update the scripts? build can be done with standard dotnet cli, and we are not planning to pack this project

no npm pack?

File: javascript/generated/DtdlOm.d.ts [hunk 1]
```diff
@@ -142,7 +142,8 @@ export type IntegerType = IntegerInfo;
 
 export interface InterfaceInfo extends EntityInfo {
   EntityKind: 'Interface';
-  contents: { [name: string]: string };
+  // NOTE: The below line was modified by hand.
+  contents: { [name: string]: EntityType | string };
```
Comment (jrdouceur) [inline javascript/generated/DtdlOm.d.ts:146]: I had thought you were going to make this declaration in some other file; this is not really accurate for the JSON model exposed from the ParseToJson() method of the .NET parser.

**Comment by rido-min** [comment]
> It is dotnet code; shouldn't it go in the dotnet folder?
yes, but it's intended to use by JS.. as it is including files from the js folder I think it belongs to that folder.

> no npm pack?
AFAIK we have not decided if we will publish this to NPM

**Comment by jrdouceur** [comment]
> > yes, but it's intended to use by JS.. as it is including files from the js folder I think it belongs to that folder.
> 
> > AFAIK we have not decided if we will publish this to NPM

Vishnu, you will not be able to make some of these changes from a fork.  I will start a new PR that copies from this one and makes the changes Rido proposes: moving JSInterop and removing the packing.  I should have a chance to get to this later in the afternoon.

File: javascript/generated/DtdlOm.d.ts (hunk 1 above)
Comment (vishnureddy17) [reply javascript/generated/DtdlOm.d.ts:146]: You're right. Will fix

**Comment by vishnureddy17** [comment]
If we're not publishing to NPM, I think that makes a _stronger_ case for keeping the NPM pack. The NPM pack will create a local tarball, which one might want since they can't install it from an online NPM registry.

File: javascript/generated/DtdlOm.d.ts (hunk 1 above)
Comment (vishnureddy17) [reply javascript/generated/DtdlOm.d.ts:146]: I'll wait for your PR first @jrdouceur

**Comment by jrdouceur** [comment]
> If we're not publishing to NPM, I think that makes a _stronger_ case for keeping the NPM pack. The NPM pack will create a local tarball, which one might want since they can't install it from an online NPM registry.

@rido-min Any response to this?  I'm holding off on packing until this is resolved.  How do you envision the interop layer being consumed?

**Comment by rido-min** [comment]
I'd start with a sample showing how to consume the ParseToJson API from WASM. Until we got more customer evidence I would not start the packaging.. (if anyone needs the tarball, the sample instructions will show them how to do it).

File: dotnet/src/JsInterop/JsInterop.csproj [hunk 2]
```diff
@@ -2,15 +2,24 @@
 
   <PropertyGroup>
     <TargetFramework>net7.0</TargetFramework>
+    <RuntimeIdentifier>browser-wasm</RuntimeIdentifier>
+    <WasmMainJSPath>..\..\..\javascript\index.js</WasmMainJSPath>
```
Comment (rido-min) [inline dotnet/src/JsInterop/JsInterop.csproj:6]: moving this csproj to the javascript folder will avoid including files from a parent directory

File: dotnet/src/JsInterop/JsInterop.csproj [hunk 3]
```diff
@@ -36,6 +45,13 @@
     <DocumentationFile>$(OutputPath)DTDLParserInterop.xml</DocumentationFile>
   </PropertyGroup>
 
+  <ItemGroup>
+    <WasmExtraFilesToDeploy Include="..\..\..\javascript\index.d.ts" />
+    <WasmExtraFilesToDeploy Include="..\..\..\javascript\DtdlErr.d.ts" />
```
Comment (rido-min) [inline dotnet/src/JsInterop/JsInterop.csproj:50]: wondering if DtdlErr.d.ts should be in the generated folder as DtdlOm

File: javascript/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+
+import { DtdlObjectModel } from "./generated/DtdlOm";
```
Comment (rido-min) [inline javascript/index.d.ts:2]: I dont see the generated folder in the produced package

**Review by rido-min — CHANGES_REQUESTED** [review]
the typings are not resolved from the package

File: dotnet/src/JsInterop/JsInterop.csproj (hunk 3 above)
Comment (jrdouceur) [reply dotnet/src/JsInterop/JsInterop.csproj:50]: DtdlErr.d.ts is not a generated file.  It is handwritten.

File: javascript/index.d.ts (hunk 4 above)
Comment (jrdouceur) [reply javascript/index.d.ts:2]: I'm guessing this will be a non-issue once the packing is removed?

File: javascript/index.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+export async function parseAsync(jsonTexts) {
+    /* Start dotnet runtime if not already started */
+    dotnetExports ??= await createRuntimeAndGetExports()
+
+    jsonTexts = typeof jsonTexts === 'string' ? [jsonTexts] : jsonTexts
+
+    let objectModelJson
+    try {
+        objectModelJson = await dotnetExports.ModelParserInterop.ParseToJsonAsync(Array.from(jsonTexts))
+    } catch (modelParsingError) {
```
Comment (jrdouceur) [inline javascript/index.js:21]: My knowledge of JS is weak, but is line 21 correct?  This is catching an exception into variable 'modelParsingError', which is then thrown below on line 29, after having its 'exceptionInfo' field set to the value of the 'message' field from a variable named 'error', which does not seem to be declared anywhere in this file.

File: javascript/index.d.ts (hunk 4 above)
Comment (vishnureddy17) [reply javascript/index.d.ts:2]: @rido-min Yep. My mistake.

File: javascript/index.js (hunk 5 above)
Comment (vishnureddy17) [reply javascript/index.js:21]: Dumb mistake on my part. Good catch.

**Comment by jrdouceur** [comment]
Superseded by PR #23

**Pull request closed without merging**